### ISSUE TITLE
Password Reset: what to do if an account can't found based on an email address

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#2030]
The Password Reset feature was developed in #416 but now that Dataverse 4.0 is in production at https://dataverse.harvard.edu we might want to reconsider some design decisions.

## What if someone attempts to reset the password of an email address that does not exist in the system?

Let's say I go to https://dataverse.harvard.edu/passwordreset.xhtml and type in "president@whitehouse.gov" and click "Submit Password Request" like this:

![account_password_reset_-_harvard_dataverse_-_2015-04-16_09 11 26](https://cloud.githubusercontent.com/assets/21006/7181525/60e20b6e-e419-11e4-94d0-e7f6f4343ee3.png)

What should happen? Right now the user will see "An email has been sent with further instructions to president@whitehouse.gov" like this:

![account_password_reset_-_harvard_dataverse_-_2015-04-16_09 11 42](https://cloud.githubusercontent.com/assets/21006/7181538/815b5bf2-e419-11e4-997c-96177da0c136.png)

What does someone looking at the server logs see? Something like this:

`Couldn't find single account using president@whitehouse.gov`

When an account can not be found based on "president@whitehouse.gov" should we send an email to president@whitehouse.gov? @eaquigley and I decided not to at https://github.com/IQSS/dataverse/issues/416#issuecomment-53076660 because it would be too spammy. So **by design** an email saying "an account associated with this email address could not be found" is **not** sent.

If we decide to send an email when the account doesn't exist, what should the email say? Here is an example from [the blog post that inspired the design of the password reset feature](http://www.troyhunt.com/2012/05/everything-you-ever-wanted-to-know.html):

![no-account](https://cloud.githubusercontent.com/assets/21006/7181638/35f62b14-e41a-11e4-84fb-2ca17d810d0a.png)

In summary, there a few things to decide about the behavior we want when the user types in an email address that can't be found in the system:

- What should we show the user?
- What (if anything) should a superuser see?
- Should an email be sent to the address that is not associate with any account in the system?

**Comment by bencomp** [comment]
Not sure about best practices for this, but the possibility to submitting email addresses for password resets should be limited. When the address cannot be found, tell the user it is not associated with an account.

The audit log should have a record of the requested password reset. A log monitoring system should notify the administrator of suspicious behaviour when there appears to be an attack going on.

Don't send an email. It's too spammy. If the email address does belong to the user, but it's not associated with the user's account, s/he can still contact the administrator.

**Comment by posixeleni** [comment]
This is how figshare handles password reset if you enter an email address that is not in their database:

![image](https://cloud.githubusercontent.com/assets/5013837/7182340/b00d5d38-e41e-11e4-8bb7-6b2e8d3cb124.png)

**Comment by pdurbin** [comment]
>  Right now the user will see "An email has been sent with further instructions to president@whitehouse.gov"...
> ...
> So by design an email saying "an account associated with this email address could not be found" is not sent.

I know those two statement sound contradictory but the intention was to **respect users' privacy**. If hackers know your email address, **should they be able to go to a site and discover if you have an account on that site by filling out the password reset form with your email address?**

Maybe it would be better to simply let the hackers (and the users!) know whether or not an account exists for the user based on the email address. As @posixeleni points out, the user will know immediately to try a different email address, which is a very good thing.

**Comment by tercer** [comment]
This just tripped me up today.  I thought my account was through this email address:

![screen shot 2015-05-05 at 11 33 00 am](https://cloud.githubusercontent.com/assets/2085037/7478497/343ddda8-f329-11e4-83f4-cd66591beca8.png)

But I had no account at that email.  So waited patiently. Like the gullible fool that I am.  

I understand the privacy issue.  For the present, until you resolve this deeply philosophical quandry, wouldn't it work to restate the message as something like:

"If this email is associated with an account, then an email will be sent with further instructions to user@some.where.edu"

**Comment by eaquigley** [comment]
We should change the message displayed in the UI to say what @tercer suggested. I think it is apparent based off log in issues we've had since 4.0 was launched that this is the right thing to do. Additionally, I agree with @bencomp, "The audit log should have a record of the requested password reset. A log monitoring system should notify the administrator of suspicious behaviour when there appears to be an attack going on."

So, to do list:

- [ ] Change text in UI to read, If this email is associated with an account, then an email will be sent with further instructions to [insert email address here]
- [ ] Have a record in the audit log of the requested password resets with the system notifying the administrator of suspicious behavior when there appears to be an attack going on.

**Comment by pdurbin** [comment]
@tercer that was a fantastic suggestion. Thanks!

@eaquigley I went ahead and pushed the wording change into master in c4a546d so it will show up in the next build. I'll move this ticket to the current milestone and into QA.

@eaquigley your other suggestion is a great idea but quite involved so I created a separate ticket for it: #2134

**Comment by kcondon** [comment]
OK, text is correct, the other issue Phil opened in a separate ticket. I am closing this.
